resharper/resharper-unity/src/Rider/UnitTesting/RunViaUnityEditorStrategy.cs

```diff
@@ -107,31 +117,77 @@ Task IUnitTestRunStrategy.Run(IUnitTestRun run)
             // https://docs.unity3d.com/ScriptReference/Compilation.CompilationPipeline-assemblyCompilationFinished.html
             // https://docs.unity3d.com/ScriptReference/Compilation.CompilationPipeline-assemblyCompilationStarted.html
             // Note that those events are only available for Unity 5.6+
-//            mySolution.Locks.ExecuteOrQueueEx(run.Lifetime, "RefreshBeforeUT", () =>
-//            {
-//                myRiderSolutionSaver.Save(run.Lifetime, mySolution, async () =>
-//                {
-//                    await myUnityRefresher.Refresh(true);    
-//                }); 
-//            });
-
-            var launch = SetupLaunch(run);
-            mySolution.Locks.ExecuteOrQueueEx(run.Lifetime, "ExecuteRunUT", () =>
+            bool working = true;
+            mySolution.Locks.ExecuteOrQueueEx(run.Lifetime, "RefreshBeforeUT", () =>
+            {
+                myRiderSolutionSaver.Save(run.Lifetime, mySolution, async () =>
+                {
+                    await myUnityRefresher.Refresh(false);
+                    working = false;
+                });
+            });
+
+            JetDispatcher.RunOrSleep(() => !(!run.Lifetime.IsAlive || !working), TimeSpan.FromMinutes(30));
+            working = true;
+
+            mySolution.Locks.ExecuteOrQueueEx(run.Lifetime, "Wait EditorState != UnityEditorState.Refresh", () =>
+            {
+                mySolution.Locks.QueueRecurring(run.Lifetime, "Periodic wait EditorState != UnityEditorState.Refresh",
+                    TimeSpan.FromSeconds(1), () =>
+                    {
+                        if (myEditorProtocol.UnityModel.Value != null)
+                        {
+                            var rdTask = myEditorProtocol.UnityModel.Value.GetUnityEditorState.Start(RdVoid.Instance);
```

Comment:
we should have extension method that you shouldn't write RdVoid.Instance

Reply:
I don't see such method.